File: Medium/2807. Insert Greatest Common Divisors in Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    public ListNode insertGreatestCommonDivisors(ListNode head) {
        ListNode now = head;
        while (now.next != null) {
            int GCD;
            if (now.val % now.next.val == 0 || now.next.val % now.val == 0) {
                GCD = Math.min(now.val, now.next.val);
            } else {
                GCD = gcd(now.val, now.next.val);
            }
            now.next = new ListNode(GCD, now.next);
            now = now.next.next;
        }
        return head;
    }
}

File: Medium/2807. Insert Greatest Common Divisors in Linked List.kt
class ListNode(var `val`: Int) {
    var next: ListNode? = null
    constructor(`val`: Int, next: ListNode?) : this(`val`) {
        this.next = next
    }
}

class Solution {
    private fun gcd(a: Int, b: Int): Int {
        var a = a
        var b = b
        while (b != 0) {
            val temp = b
            b = a % b
            a = temp
        }
        return a
    }

    fun insertGreatestCommonDivisors(head: ListNode?): ListNode? {
        var dummy = head
        while (dummy?.next != null) {
            dummy.next = ListNode(gcd(dummy.`val`, dummy.next.`val`), dummy.next)
            dummy = dummy.next.next
        }
        return head
    }
}

